ansible/roles/test/files/ptftests/inner_hash_test.py

```diff
@@ -0,0 +1,505 @@
+            else:
+                self.check_balancing(self.next_hop.get_next_hop(), hit_count_map)
+
+
+    def check_ip_route(self, hash_key, outer_encap_format, src_port, ip_src, ip_dst, sport, dport, ip_proto):
+        if (outer_encap_format == 'vxlan'):
+            (matched_index, received) = self.check_ip_route_vxlan(hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+        elif (outer_encap_format == 'nvgre'):
+            (matched_index, received) = self.check_ip_route_nvgre(hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+        else:
+            raise Exception("Logic issue, unexpected outer encap format {}".format(outer_encap_format))
+
+        matched_port = self.exp_port_list[matched_index]
+        return (matched_port, received)
+
+
+    def _get_ip_proto(self, ipv6=False):
+        # ip_proto 2 is IGMP, should not be forwarded by router
+        # ip_proto 254 is experimental
+        # MLNX ASIC can't forward ip_proto 254, BRCM is OK, skip for all for simplicity
+        skip_ports = [2, 254]
+        if ipv6:
+            # Skip ip_proto 0 for IPv6
+            skip_ports.append(0)
+
+        while True:
+            ip_proto = random.randint(0, 255)
+            if ip_proto not in skip_ports:
+                return ip_proto
+
+
+    def generate_inner_pkt(self, sport, dport, ip_src, ip_dst, ip_proto):
+        rand_int = random.randint(1, 99)
+        src_mac = '00:12:ab:34:cd:' + str(rand_int)
+        dst_mac = str(rand_int) + ':12:ab:34:cd:00'
+        pkt = simple_tcp_packet(
+                        eth_dst=dst_mac,
+                        eth_src=src_mac,
+                        ip_src=ip_src,
+                        ip_dst=ip_dst,
+                        tcp_sport=sport,
+                        tcp_dport=dport,
+                        ip_ttl=64)
+
+        pkt['IP'].proto = ip_proto
+        return pkt
+
+
+    def check_ip_route_nvgre(self, hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto):
+        '''
+        @summary: Check nvgre based inner packet hashing works
+        @param hash_key: hash_key we are varying in this iteration
+        @param src_port: index of port to use for sending packet to switch
+        @param ip_dst: Inner packet Destination IP address
+        @param ip_src: Inner packet Source IP address
+        @param dport: Inner packet dst port
+        @param sport: Inner packet src port
+        @param ip_proto: Inner packet ip protocol
+        '''
+        if ip_network(unicode(self.outer_dst_ip)).version == 4:
+            (matched_index, received) = self.check_ipv4_route_nvgre(hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+        else:
+            (matched_index, received) = self.check_ipv6_route_nvgre(hash_key, src_port, ip_dst, ip_src, dport, sport, ip_proto)
+
+        assert received
+
+        logging.info("Received packet at " + str(matched_index))
+        time.sleep(0.02)
```

Comment:
@anish-n why do we need this?

Reply:
This part of the logic is identical to existing hash test. https://github.com/Azure/sonic-mgmt/blob/master/ansible/roles/test/files/ptftests/hash_test.py#L158 . I want to maintain logic from existing hash test as and where applicable to the inner hash test.